src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceBatchSink.java

```diff
@@ -0,0 +1,137 @@
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Plugin inserts records into Salesforce using Salesforce Bulk API.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name(SalesforceBatchSink.PLUGIN_NAME)
+@Description("Writes records to Salesforce")
+public class SalesforceBatchSink extends BatchSink<StructuredRecord, NullWritable, CSVRecord> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceBatchSink.class);
+
+  public static final String PLUGIN_NAME = "Salesforce";
+
+  private final SalesforceSinkConfig config;
+
+  public SalesforceBatchSink(SalesforceSinkConfig config) throws ConnectionException {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate(pipelineConfigurer.getStageConfigurer().getInputSchema());
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) {
+    Schema inputSchema = context.getInputSchema();
+    config.validate(inputSchema);
+
+    context.addOutput(Output.of(config.referenceName, new SalesforceOutputFormatProvider(config)));
+
+    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
+    lineageRecorder.createExternalDataset(inputSchema);
+    // Record the field level WriteOperation
+    if (inputSchema.getFields() != null && !inputSchema.getFields().isEmpty()) {
+      lineageRecorder.recordWrite("Write", "Wrote to Salesforce sink",
+                                  inputSchema.getFields().stream()
+                                    .map(Schema.Field::getName)
+                                    .collect(Collectors.toList()));
+    }
+  }
+
+  @Override
+  public void transform(StructuredRecord record, Emitter<KeyValue<NullWritable, CSVRecord>> emitter) {
+    List<String> fieldNames = new ArrayList<>();
+    List<Object> values = new ArrayList<>();
+
+    for (Schema.Field field : record.getSchema().getFields()) {
+      String fieldName = field.getName();
+      Object value = convert(record.get(fieldName), field);
+
+      fieldNames.add(fieldName);
+      values.add(value);
+    }
+
+    CSVRecord csvRecord = new CSVRecord(fieldNames, values);
+    emitter.emit(new KeyValue<>(null, csvRecord));
+  }
+
+  private Object convert(Object value, Schema.Field field) {
+    // don't convert null to avoid NPE
+    if (value == null) {
+      return null;
+    }
+
+    Schema fieldSchema = field.getSchema();
+
+    if (fieldSchema.isNullable()) {
+      fieldSchema = fieldSchema.getNonNullable();
+    }
+
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      Instant instant;
+      switch (logicalType) {
+        case DATE:
+          // convert epoch day to yyyy-mm-dd format
+          return LocalDate.ofEpochDay((Integer) value).toString();
+        case TIMESTAMP_MICROS:
```

Comment:
Well the approach is the same as in other Salesforce plugins. We had handling of millis some time ago, but Arina removed it explaining the reasoning at https://github.com/data-integrations/salesforce/pull/14#discussion_r275868661

Reply:
It's true that the UI currently only understands micros (this will likely eventually change) so it makes sense for the source to output micros. However, the sink can take input from any other plugin, and those plugins might use millis. Although we likely won't see this scenario with the current set of plugins, it's still better to handle it in case it gets more prevalent in the future, especially since handling it seems straightforward.

Reply:
done